### PR TITLE
PRMP-198_199 - Lambdas: DataCollectionLambda and StatisticalReportLambda

### PR DESCRIPTION
**Pull request by joefong-nhs**

### CODE REVIEW

File: .github/workflows/base-lambdas-reusable-deploy-all.yml [hunk 1]
```diff
@@ -316,4 +316,32 @@ jobs:
...
+    name: Deploy data collection lambda
+    uses: ./.github/workflows/base-lambdas-reusable-deploy.yml
+    with:
+      environment: ${{ inputs.environment}}
+      python_version: ${{ inputs.python_version }}
+      build_branch: ${{ inputs.build_branch}}
+      sandbox: ${{ inputs.sandbox }}
+      lambda_handler_name: data_collection_handler
+      lambda_aws_name: DataCollectionLambda
+      lambda_layer_names: "core_lambda_layer,data_lambda_layer"
```
Comment (abbas-khan10) [inline .github/workflows/base-lambdas-reusable-deploy-all.yml:331]: Do both these lambdas require polars?

File: lambdas/models/cloudwatch_logs_query.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
+from dataclasses import dataclass
```
Comment (abbas-khan10) [inline lambdas/models/cloudwatch_logs_query.py:1]: I wonder if this belongs in the utils folder alongside the dynamo filters, as this is not really a models file?

File: .github/workflows/base-lambdas-reusable-deploy-all.yml (hunk 1 above)
Comment (joefong-nhs) [reply .github/workflows/base-lambdas-reusable-deploy-all.yml:331]: Yes, I have used polars in the DataCollection lambda for a few metrics that are a bit tricky and need some calculation, for example, "average number of records per patient per each ODS code"

File: lambdas/models/cloudwatch_logs_query.py (hunk 2 above)
Comment (joefong-nhs) [reply lambdas/models/cloudwatch_logs_query.py:1]: Thanks! Agree that this makes more sense to be in the utils folder. I will move them now.

File: lambdas/services/data_collection_service.py [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+
+        record_store_data = self.get_record_store_data(
+            dynamodb_scan_result, s3_list_objects_result
+        )
+        organisation_data = self.get_organisation_data(dynamodb_scan_result)
+        application_data = self.get_application_data()
+
+        return record_store_data + organisation_data + application_data
+
+    def write_to_local_dynamodb_table(self, all_statistic_data: list[StatisticData]):
```
Comment (abbas-khan10) [inline lambdas/services/data_collection_service.py:73]: Why have we included 'local' here?

File: lambdas/services/data_collection_service.py [hunk 4]
```diff
@@ -0,0 +1,324 @@
...
+        self.dynamodb_service.batch_writing(
+            table_name=self.output_table_name, item_list=item_list
+        )
+
+        logger.info("Finish writing all data to dynamodb table")
+
+    def scan_dynamodb_tables(self) -> list[dict]:
+        all_results = []
+
+        project_expression = "CurrentGpOds,NhsNumber,FileLocation,ContentType"
```
Comment (abbas-khan10) [inline lambdas/services/data_collection_service.py:88]: Should have have these hardcoded here? Perhaps a string interpolation of our enum could be a better option?

File: lambdas/services/data_collection_service.py [hunk 5]
```diff
@@ -0,0 +1,324 @@
...
+            ods_code = item["CurrentGpOds"]
+            file_type = item["ContentType"]
+            file_types_grouped_by_ods_code[ods_code].add(file_type)
+
+        return [
+            {"ods_code": ods_code, "number_of_document_types": len(file_types)}
+            for ods_code, file_types in file_types_grouped_by_ods_code.items()
+        ]
+
+    def get_average_number_of_file_per_patient(
```
Comment (abbas-khan10) [inline lambdas/services/data_collection_service.py:295]: *files

File: lambdas/services/data_collection_service.py (hunk 3 above)
Comment (joefong-nhs) [reply lambdas/services/data_collection_service.py]: sorry, this was a legacy name when I built this with a locally hosted dynamodb
renamed to `write_to_dynamodb_table` now

File: lambdas/services/data_collection_service.py (hunk 4 above)
Comment (joefong-nhs) [reply lambdas/services/data_collection_service.py]: Thanks, totally agree that we should use our enum instead.
I have replaced this line and all other hardcoded field names of DocumentReference in this class with values from enum

File: lambdas/services/data_collection_service.py (hunk 5 above)
Comment (joefong-nhs) [reply lambdas/services/data_collection_service.py]: Nice catch! :>  Fixed just now

**Review by abbas-khan10 — APPROVED** [review]

**Pull request merged**